### ISSUE TITLE
False positive `no-matching-overload` on async method overload

### ISSUE THREAD

**Issue by Apakottur** [astral-sh/ty#3825]
### Summary

Consider the following code:
```python
# a.py

from typing import Any, Generic, TypeVar, overload


_Single = TypeVar("_Single")

_Multiple = TypeVar("_Multiple", bound=tuple[Any, Any, *tuple[Any, ...]])

_TP = TypeVar("_TP", bound=tuple[Any, ...])


class Select(Generic[_TP]):
    pass


@overload
async def _query(query: Select[tuple[_Single]]) -> list[_Single]: ...


@overload
async def _query(query: Select[_Multiple]) -> list[_Multiple]: ...


async def _query(query: Select[tuple[_Single]] | Select[_Multiple]) -> list[_Single] | list[_Multiple]:
    raise NotImplementedError


async def query(query: Select[tuple[_Single]] | Select[_Multiple]) -> list[_Single] | list[_Multiple]:
    return await _query(query)
```

Running `ty check a.py` gives:
```text
error[no-matching-overload]: No overload of function `_query` matches arguments
  --> a.py:30:18
   |
30 |     return await _query(query)
   |                  ^^^^^^^^^^^^^
   |
info: First overload defined here
  --> a.py:17:1
   |
17 | / @overload
18 | | async def _query(query: Select[tuple[_Single]]) -> list[_Single]: ...
   | |_____________________________________________________________________^ First overload defined here
   |
info: Possible overloads for function `_query`:
info:   [_Single](query: Select[tuple[_Single]]) -> CoroutineType[Any, Any, list[_Single]]
info:   [_Multiple](query: Select[_Multiple]) -> CoroutineType[Any, Any, list[_Multiple]]
info: Overload implementation defined here
  --> a.py:25:11
   |
25 | async def _query(query: Select[tuple[_Single]] | Select[_Multiple]) -> list[_Single] | list[_Multiple]:
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |

Found 1 diagnostic
```

mypy and pyright accept the code without any issues.

Is this possibly the same underlying issue as https://github.com/astral-sh/ty/issues/3675 and https://github.com/astral-sh/ty/issues/541?

### Version

0.0.51

**Comment by carljm** [comment]
This is a symptom of the TODO at https://github.com/astral-sh/ruff/blob/b2e333fe40f8e11a81ba6f6ea7bd062879612bc8/crates/ty_python_semantic/src/types/call/bind.rs#L4876-L4884 -- I think it might be resolved (or at least unblocked) by #2521
